### PR TITLE
Fix help, add RPC tests for getblockheader

### PR DESCRIPTION
**Pull request by jamesob**
Continuing to chip away at the untested RPC interface. This PR:
- adds tests for `getblockheader` (previously uncovered)
- adds missing documentation to the `getblockheader` RPC help
- adds two assertion utilities to the RPC testing framework

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Nice. utACK.
tiny nit: don't label rpcblockchain.cpp changes (help output) as [test]. Maybe better open a 2nd PR.

File: qa/rpc-tests/test_framework/util.py [hunk 1]
```diff
@@ -407,5 +407,22 @@ def assert_raises(exc, fun, *args, **kwds):
     else:
         raise AssertionError("No exception raised")
 
+def assert_is_hex_string(string):
+    try:
+        int(string, 16)
+    except Exception as e:
+        raise AssertionError(
+            "Couldn't interpret %r as hexidecimal; raised: %s" % (string, e))
```
Comment (laanwj) [inline qa/rpc-tests/test_framework/util.py]: hexadecimal

Comment (jamesob) [inline qa/rpc-tests/test_framework/util.py]: fixed

**Comment by jamesob** [comment]
@jonasschnelli good point, thanks. I've changed the PR title to reflect this.

**Comment by morcos** [comment]
utACK modulo question

**Comment by jamesob** [comment]
talked with @morcos out-of-band, sounds as though he's cool with changes as they are.

File: src/rpcblockchain.cpp [hunk 2]
```diff
@@ -323,7 +323,8 @@ UniValue getblockheader(const UniValue& params, bool fHelp)
             "  \"bits\" : \"1d00ffff\", (string) The bits\n"
             "  \"difficulty\" : x.xxx,  (numeric) The difficulty\n"
             "  \"previousblockhash\" : \"hash\",  (string) The hash of the previous block\n"
-            "  \"nextblockhash\" : \"hash\"       (string) The hash of the next block\n"
+            "  \"nextblockhash\" : \"hash\",      (string) The hash of the next block\n"
+            "  \"chainwork\" : \"0000...1f3\"     (string) Number of hashes expected to produce the current chain (in hex)\n"
```
Comment (petertodd) [inline src/rpcblockchain.cpp]: What about "Expected number of hashes required to produce the current chain"?

Current wording could be read as "do that work, will get the current chain", which is a bit unlikely. :)

**Comment by petertodd** [comment]
utACK modulo wording

File: src/rpcblockchain.cpp (hunk 2 above)
Comment (jamesob) [inline src/rpcblockchain.cpp]: sure thing. fixed

File: qa/rpc-tests/test_framework/util.py [hunk 3]
```diff
@@ -407,5 +407,22 @@ def assert_raises(exc, fun, *args, **kwds):
...
+        raise AssertionError(
+            "Couldn't interpret %r as hexadecimal; raised: %s" % (string, e))
+
+def assert_is_hash_string(string, length=64):
+    if not isinstance(string, basestring):
+        raise AssertionError("Expected a string, got type %r" % type(string))
+    elif length and len(string) != length:
+        raise AssertionError(
+            "String of length %d expected; got %d" % (length, len(string)))
+    elif not re.match('[abcdef0-9]+', string):
```
Comment (laanwj) [inline qa/rpc-tests/test_framework/util.py]: I think you want:
```
    elif not re.match('[abcdef0-9]+$', string):
```

Comment (jamesob) [inline qa/rpc-tests/test_framework/util.py]: Very true :). willfix

Comment (jamesob) [inline qa/rpc-tests/test_framework/util.py]: fixed

File: src/rpcblockchain.cpp [hunk 4]
```diff
@@ -323,7 +323,8 @@ UniValue getblockheader(const UniValue& params, bool fHelp)
             "  \"bits\" : \"1d00ffff\", (string) The bits\n"
             "  \"difficulty\" : x.xxx,  (numeric) The difficulty\n"
             "  \"previousblockhash\" : \"hash\",  (string) The hash of the previous block\n"
-            "  \"nextblockhash\" : \"hash\"       (string) The hash of the next block\n"
+            "  \"nextblockhash\" : \"hash\",      (string) The hash of the next block\n"
+            "  \"chainwork\" : \"0000...1f3\"     (string) Expected number of hashes required to produce the current chain (in hex)\n"
```
Comment (instagibbs) [inline src/rpcblockchain.cpp]: It's the expected number of hashes up to *this point* in the chain, not the entire chain. AFAIK

Comment (instagibbs) [inline src/rpcblockchain.cpp]: I added a similar line for getblock at #7232

**Comment by jamesob** [comment]
ping on this

**Comment by sipa** [comment]
I'll start reviewing again after new year.

File: src/rpcblockchain.cpp (hunk 4 above)
Comment (jamesob) [inline src/rpcblockchain.cpp]: This message is based on [@sipa's description of chainwork](https://bitcoin.stackexchange.com/questions/26869/what-is-chainwork) and [was vetted previously by @petertodd](https://github.com/bitcoin/bitcoin/pull/7194#discussion_r47325775). You may want to close the PR you reference above; it was opened after this one and doesn't include tests.

Comment (instagibbs) [inline src/rpcblockchain.cpp]: Ok, I disagree that it's clear, but if everyone else is fine with it I can conform as well.

Regardless, my pull is not covered here, so not sure why I'd close it?

Comment (jamesob) [inline src/rpcblockchain.cpp]: @instagibbs thanks!

Comment (instagibbs) [inline src/rpcblockchain.cpp]: Perhaps my newer suggestion is equitable? 

https://github.com/bitcoin/bitcoin/pull/7232/files#r48769375

**Comment by laanwj** [comment]
utACK

**Pull request merged**